### PR TITLE
:seedling: Add additional blocks to release note generation

### LINKED ISSUE #9248: Add additional headers to the release notes tool output
As mentioned in https://github.com/kubernetes-sigs/cluster-api/issues/9104,

In recent releases, there are some additional header sections that have been manually generated, such as "Highlights" and "Deprecation Warning" - these should be automatically generated by the release notes tool even if the content is not automated, a template with placeholders is fine.

Release notes tool: https://github.com/kubernetes-sigs/cluster-api/blob/main/hack/tools/release/notes.go

### PR DESCRIPTION
**Pull request by mjlshen**
**What this PR does / why we need it**:

This PR allows the existing release notes generation tool to generate (optionally in some cases) commonly used headers that have been manually filled out in the past:

* An optional header when generating a release candidate
* A highlights header with filler to manually replace
* An optional header about deprecation warnings

**Which issue(s) this PR fixes**:

Fixes #9248
Part of #9104

---

Sample header output:
```
❯ go run hack/tools/release/notes.go -from release-1.5 -workers 1 -release-candidate -deprecation
🚨 This is a RELEASE CANDIDATE. Use it only for testing purposes. If you find any bugs, file an [issue](https://github.com/kubernetes-sigs/cluster-api/issues/new).
## 👌 Kubernetes version support

- Management Cluster: v1.**X**.x -> v1.**X**.x
- Workload Cluster: v1.**X**.x -> v1.**X**.x

[More information about version support can be found here](https://cluster-api.sigs.k8s.io/reference/versions.html)

## Highlights

* REPLACE ME

## Deprecation Warning

REPLACE ME: A couple sentences describing the deprecation, including links to docs.

* [GitHub issue #REPLACE ME](REPLACE ME)

## Changes since release-1.5
```

/area release

### CODE REVIEW

File: hack/tools/release/notes.go
Comment (mjlshen) [inline hack/tools/release/notes.go:311]: This comment didn't make sense to me, based on the [original intention](https://github.com/kubernetes-sigs/cluster-api/commit/d76ea10ffbc9ab697f56ece40ec6b67ca252f423#diff-818b0767917bf0415c71e90e83d21b6faf07a4096264d71de9646f2c1ab4f48aR123-R124) it seemed to be better placed with `## Changes since`, so I moved it

**Comment by mjlshen** [comment]
cc @kubernetes-sigs/cluster-api-release-team

File: hack/tools/release/notes.go
Comment (killianmuldoon) [inline hack/tools/release/notes.go:313]: Do we need to do this for a beta aswell? Would it be easier to detect this from the structure of the release tag rather than changing the flags depending on the kind of release?

**Review by killianmuldoon — COMMENTED** [review]
Looks generally good - one overall comment.

As this gets more complex it might be a good idea to move to Go templates. That way the information to fill out the additional headers can be maintained through the release cycle.

File: hack/tools/release/notes.go
Comment (g-gaston) [inline hack/tools/release/notes.go:77]: I would prefer if this was true by default] and when not needed, we can just remove the section from the notes.
The idea behind these changes is to prevent us from forgetting something in the notes. If the section is there and we don't have content, we won't miss deleting it. However, it's possible that we might miss passing this flag.

File: hack/tools/release/notes.go
Comment (g-gaston) [reply hack/tools/release/notes.go:311]: I think we can remove it? the `repo` variable contains this info I believe

File: hack/tools/release/notes.go
Comment (g-gaston) [reply hack/tools/release/notes.go:313]: Indeed, this would be a good idea. It could also help with other "variable" sections, like the expandable blocks for RC/Beta (with the full commits diff against las minor release).

However, it requires re-designing the input to this command, at least partly. Right now, it takes the base tag/commit you want to use to compute the diff with HEAD (of current branch). This approach would move it to take the "next tag" as the input and it will need to compute the `fromTag` dynamically based on certain rules (which I think we can easily come up with based on our release schedule).

I'll leave that up to y'all, I don't have a hard opinion, but I would be ok merging this as is (we get some value delivered immediately) and then working on this input redesign as one of our next improvement tasks.

File: hack/tools/release/notes.go
Comment (mjlshen) [reply hack/tools/release/notes.go:77]: That's fair - do you think there's even value to have the flag then vs just always templating this out?

File: hack/tools/release/notes.go
Comment (furkatgofurov7) [reply hack/tools/release/notes.go:313]: if we need beta as well, then renaming it to something like more generic, i.e `pre-release version` or similar might fit better

File: hack/tools/release/notes.go
Comment (furkatgofurov7) [reply hack/tools/release/notes.go:311]: +1 to remove it

**Review by furkatgofurov7 — COMMENTED** [review]
Overall lgtm, pending to clarify few open questions from reviewers and should be good to take it in! Thanks @mjlshen for working on this

**Comment by richardcase** [comment]
> As this gets more complex it might be a good idea to move to Go templates. That way the information to fill out the additional headers can be maintained through the release cycle.

I like the idea of using a go template. This would be similar to the kubernetes release-notes tool that uses go templates (see [here](https://github.com/kubernetes/release/blob/master/pkg/notes/document/template.go#L29)). It also allows you to [override the template](https://github.com/kubernetes/release/blob/master/cmd/release-notes/main.go#L166-L177)

File: hack/tools/release/notes.go
Comment (richardcase) [inline hack/tools/release/notes.go:77]: Comment says the default is **false** but the flag definition has it as **true**

```suggestion
	deprecation                 = flag.Bool("deprecation", false, "If enabled, will add a templated deprecation warning header. (default false)")
```

File: hack/tools/release/notes.go
Comment (mjlshen) [reply hack/tools/release/notes.go:313]: I think we can do this easily if we switch to the kubernetes release tool, which I'm taking a look at through #9249

As it is now I don't think there's a straightforward way to make this "smarter" - I can update this PR to use the naming `--pre-release-version` though since I agree that's clearer.

**Comment by mjlshen** [comment]
I took a look at switching over to the kubernetes release tool here: https://github.com/kubernetes-sigs/cluster-api/issues/8292#issuecomment-1692656784

I think it's a bigger effort, so I'm leaning towards taking the easy win here and continuing discussion in the issue to see where we want to go next with this release notes tool (bigger refactor or changes upstream + switching over).

**Comment by furkatgofurov7** [comment]
> I took a look at switching over to the kubernetes release tool here: [#8292 (comment)](https://github.com/kubernetes-sigs/cluster-api/issues/8292#issuecomment-1692656784)
> 
> I think it's a bigger effort, so I'm leaning towards taking the easy win here and continuing discussion in the issue to see where we want to go next with this release notes tool (bigger refactor or changes upstream + switching over).

Agree, currently, the best way forward for us should be to take this in and continue the switching over to the new process in the original issue. 

/lgtm

**Comment by richardcase** [comment]
> I think it's a bigger effort, so I'm leaning towards taking the easy win here and continuing discussion in the issue to see where we want to go next with this release notes tool (bigger refactor or changes upstream + switching over).

Sorry I wasn't suggesting that you adopt it, just pointing out that i think its a good idea to adopt a go template (as other tools have done).

**Comment by mjlshen** [comment]
Oh no apology needed, it was a good suggestion! I was just sharing my initial findings/conclusion in case I got something wrong too

File: hack/tools/release/notes.go
Comment (vincepri) [inline hack/tools/release/notes.go:76]: Is this being set locally? Would there be a way to understand which tag we're running against and set this flag automatically? For example: v1.0.0-beta.1 I'd expect the same behavior, and similar for alpha.X or rc.X

Comment (mjlshen) [reply hack/tools/release/notes.go:76]: Ah we talked about this here: https://github.com/kubernetes-sigs/cluster-api/pull/9247#discussion_r1300079613 it's doable long-term with a bigger refactor, but we're mostly just trying to make sure we have the standardized formatting somewhere

File: hack/tools/release/notes.go
Comment (sbueringer) [inline hack/tools/release/notes.go:313]: Can we use the repo variable here? Then it would also work in other repos

File: hack/tools/release/notes.go
Comment (sbueringer) [inline hack/tools/release/notes.go:77]: Not sure how often we will use this. Is the plan to use it for other things apart from getting rid of v1alpha3/v1alpha4? Maybe the default should be false?

File: hack/tools/release/notes.go
Comment (sbueringer) [inline hack/tools/release/notes.go:74]: No idea why but in some cases apparently the flags package already adds the default in other cases it doesn't?

```bash
 go run ./hack/tools/release/notes.go --help
Usage of /var/folders/v0/mny6w2dd33gfzwmwxb3wyyx40000gq/T/go-build4290625525/b001/exe/notes:
  -add-kubernetes-version-support
        If enabled, will add the Kubernetes version support header. (default true)
  -deprecation
        If enabled, will add a templated deprecation warning header. (default true) (default true)
  -from string
        The tag or commit to start from.
  -pre-release-version
        If enabled, will add a pre-release warning header. (default false)
  -prefix-area-label
        If enabled, will prefix the area label. (default true)
  -repository string
        The repo to run the tool from. (default "kubernetes-sigs/cluster-api")
  -since string
        Include commits starting from and including this date. Accepts format: YYYY-MM-DD
  -until string
        Include commits up to and including this date. Accepts format: YYYY-MM-DD
  -workers int
        Number of concurrent routines to process PR entries. If running into GitHub rate limiting, use 1. (default 10)

```

**Review by sbueringer — COMMENTED** [review]
Just a few nits

File: hack/tools/release/notes.go
Comment (mjlshen) [reply hack/tools/release/notes.go:77]: I think this was a UX preference discussed here: https://github.com/kubernetes-sigs/cluster-api/pull/9247#discussion_r1300216670 where it's probably easier to delete this section when it's not needed vs remembering to pass the flag - I'm open to either default though

Comment (sbueringer) [reply hack/tools/release/notes.go]: Alright. Let's keep it as is

**Review by willie-yao — APPROVED** [review]
/lgtm

**Comment by sbueringer** [comment]
Thx

/approve

**Pull request merged**
